### ISSUE TITLE
`ParamSpec` destructuring via `Concatenate` fails in `__get__` overload when reusing class-level type parameters

### ISSUE THREAD

**Issue by ringohoffman** [microsoft/pyright#11558]
## Bug Description

When a generic descriptor class `Wrapper(Generic[P, R])` implements `__get__` overloads to strip `self` from a method's ParamSpec via `Concatenate`, Pyright fails to match the `overload` if the `self`-type annotation **reuses the class-level ParamSpec `P`**. Using **fresh type variables** (`Q`, `T`) in the same position works correctly.

This means the following two semantically equivalent overloads produce different results:

```python
# ❌ FAILS: reuses class-level P and R
def __get__(
    self: Wrapper[Concatenate[Any, P], R],
    obj: object,
    ...
) -> Callable[P, R]: ...

# ✅ WORKS: fresh type variables Q and T
def __get__(
    self: Wrapper[Concatenate[Any, Q], T],
    obj: object,
    ...
) -> Callable[Q, T]: ...
```

## Minimal Reproduction

[pyright playground](https://pyright-play.net/?code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwChGd9CpFSBjJMMAGyahIUJAE9cWUgHNseAkQAUjKKqgBBUmIA0UAMLF%2B-YgCN%2BaXXrpdqaUrd0BxO2hBYu2lWoAKxEMQgAZVw0DyhAtH5gXQAVCTQANT9dMAA3V34wYgATTwBKZm8oAF4oX38gkK5FACJvGoKAJRKoOJCkkFrGhuZGAGIoQABSYagAMXUASQAZQIAuKEa0BHo0eiguY3p6AFpzdP4yklJshaHh5g3iLagAdX9cEJAYReW0RWdSV3cAbW9dRoAunlZl5VNk0MBYDApFgkHBFCsorpgKR5gYjKZzL9-kCoNsAHxQABydDQILUFKgiOAADoUS0UcwKeDIXAbEZ4dTdAAqPzSejzbw0vn0HncgDWAHcRYKaVKRXk8YTGuTKap0CgQKQqZFaSjFLyQPyxfKjfQCqCoAABNIZLLZS0sqHSNBwmAI3UpEwAK3mJM%2BXu94hC82DaCgAB9iaSWv60IqCeFdfMaanLTb0iBMjlHRDna74ZaKdT5ndiA9XM8litvlZuLZ7Eg0N9NDoygCcbooAN0K81pdrn8FkW1GAffMx97QkhPGrVJOw6H4pHo59Y6TZ2oE4T0cYzM2h4CU2mKRm7TnmXm4C63R6kVBJxOfdPA4vRMuo3H159t-pDHusVTGkcWPGlc1ZGAb05T0Hx9V94mKONgVUAY3ywaQKHQb5bRANxwQBEdVCwSFJ2wNY41VOc1A1VBtWpQioBorUOAQbheAEegaVwPwaEMO9olg70LX6M5BluAB5RoAGk5jGdB6AAC3fEIoFSPwsExVYoAARSOE4YlE5hdNKcoAmCUJam0noDNKNpEj8WoYh6RgBzWMsKyeUZ5IU94XDcLhsQWIFKKgJ04BhW8uQ4VE-wxfdAsBX8KIY6k6W1UpGXAqF2X4aD70NflZRFE1pTNWVTX5X8VQYpi6N1NKDWKqBuQlUrKqZNQzyze0suvAt3Six9VwsQS3zDFcv1KJClSTKJQPTHDswdS8IKg90UuTW57keGAvNWBTa2sBtbBbLRdG0jtWku7sOG8nTYgYobJxfR6fTGj9hu-CxLV-XdNO%2BbTYgBebT0WnqVvzSKYKe58eHgkNlPDT8Yym0lfv-f6gJAqAgN6yD%2Bv4wN4bQRC0fmG60IwghmxwvC0AIqjiMEsjhpCqjap1KIatdWiWLYvhBC4niNP4QnBOEgZhjEmJViIKXmCdSV7hgXsVkUFE0Qx%2BLD1xRN3J2l4ax1kKOf1ytDbeFELUV5XQH29WYr%2B7WcV-M3PO8hLgctU3tsrPbFIdi1XKgABZMQ9E2egq1eEKrSV8sVerNAsogV0FLAbIxYAD3mKQkF-EwBbZ6ieeYmIQAQZOXMj0Pw8j3bvNj%2BPcHgbyU7TjPs9z0h85mwuBGL9VS%2B1cvK%2BYdB0kMGAw0UMOI6uKOLcUPIaVTpB0%2ByRUbsAGXIoEUICC4FxgJ7QKeZ7n%2Bv-Z8le143reBkAUHI95z7Ae8PgRGCAA)

```python
from __future__ import annotations

import functools
from typing import (
    Any, Callable, Concatenate, Generic,
    ParamSpec, Self, TypeVar, overload,
)

P = ParamSpec("P")
R = TypeVar("R")


# ─── FAILS: Reuses class-level P and R ───

class Wrapper_Reuse(Generic[P, R]):
    def __init__(self, fn: Callable[P, R]) -> None:
        self.fn = fn

    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R:
        return self.fn(*args, **kwargs)

    @overload
    def __get__(self, obj: None, objtype: type | None = None) -> Self: ...
    @overload
    def __get__(
        self: Wrapper_Reuse[Concatenate[Any, P], R],  # reuses class P, R
        obj: object,
        objtype: type | None = None,
    ) -> Callable[P, R]: ...
    @overload
    def __get__(self, obj: object, objtype: type | None = None) -> Callable[..., R]: ...
    def __get__(self, obj, objtype=None):  # type: ignore[override]
        if obj is None:
            return self
        return functools.partial(self, obj)


# ─── WORKS: Fresh type variables Q and T ───

Q = ParamSpec("Q")
T = TypeVar("T")

class Wrapper_Fresh(Generic[P, R]):
    def __init__(self, fn: Callable[P, R]) -> None:
        self.fn = fn

    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R:
        return self.fn(*args, **kwargs)

    @overload
    def __get__(self, obj: None, objtype: type | None = None) -> Self: ...
    @overload
    def __get__(
        self: Wrapper_Fresh[Concatenate[Any, Q], T],  # fresh Q, T
        obj: object,
        objtype: type | None = None,
    ) -> Callable[Q, T]: ...
    @overload
    def __get__(self, obj: object, objtype: type | None = None) -> Callable[..., R]: ...
    def __get__(self, obj, objtype=None):  # type: ignore[override]
        if obj is None:
            return self
        return functools.partial(self, obj)


# ─── Test ───

def wrap_reuse(fn: Callable[P, R]) -> Wrapper_Reuse[P, R]:
    return Wrapper_Reuse(fn)

def wrap_fresh(fn: Callable[P, R]) -> Wrapper_Fresh[P, R]:
    return Wrapper_Fresh(fn)

class MyClass_Reuse:
    @wrap_reuse
    def method(self, x: int) -> bool:
        return True

class MyClass_Fresh:
    @wrap_fresh
    def method(self, x: int) -> bool:
        return True

reveal_type(MyClass_Reuse().method)  # ❌ (...) -> bool
reveal_type(MyClass_Fresh().method)  # ✅ (x: int) -> bool
```

## Pyright Output

```
Type of "MyClass_Reuse().method" is "(...) -> bool"    ← ❌ Expected: (x: int) -> bool
Type of "MyClass_Fresh().method" is "(x: int) -> bool" ← ✅ Correct
0 errors, 0 warnings, 2 informations
```

## Expected Behavior

Both `Wrapper_Reuse` and `Wrapper_Fresh` should produce the same result: `(x: int) -> bool`.

When Pyright evaluates `MyClass_Reuse().method`, the concrete type of `self` in `__get__` is:

```
Wrapper_Reuse[(self: MyClass_Reuse, x: int), bool]
```

The overload `self: Wrapper_Reuse[Concatenate[Any, P], R]` should destructure this as:
- `Any` ← `self: MyClass_Reuse`
- `P` ← `(x: int)`
- `R` ← `bool`

This is semantically identical to what the `Q`/`T` variant does successfully.

## 3.12+ Inline Syntax

Using Python 3.12+ inline type parameter syntax on the `__get__` method works correctly, but with the error `TypeVar "P" is already in use by an outer scope`:

[pyright playground](https://pyright-play.net/?code=JYWwDg9gTgLgBAMwK4DsDGMIQDYGcBQCUEIcMAnmMCgOZyiSxwCCK5ANHAMICG22PAEbYAppy4R0PGCJTSxcAAo8oPEAGUwItJ3UjsCThABuIqNgg8AJvltoBuXHADqqsFqgBtAFTfFnACUAXQAufDgIuCsRBDgAfTjqYBgEgApcfUNEFBDuPgFhEU9-OGCASjgAWgA%2BOAA5SREwyJa4DIMAOgQUOABebNsW6NiEtHy09qzvFRpcXMUOmdxOXwBrAHcl%2BY6NpYqa0ubWiKgRGCQoHsmulFTpqFmV712H3DLByIABEzMLa3DIsN4nEaGcJpkjIIAFa5BooBQQaEULS5ZEiOAAH3qjT62Ph%2B1qegMuQ6pI%2BEW%2BpnMlhsQxiwNBKTiPm8AEVOAAVIKpAHHSa5Vw8dxmTwSKQyOQyTysDhwVlBTkK3mtREwuCq7QwdjKlqqtGoyjorFw9H9E3aloEvL8ISiTzsuBcklk5VAhKM8EGSFQ71o3omsq5OAAYjIhtywBoKGgIh1kWAsVV9CcJqOxxap3OlzamTjJzOFx6yHQmBwuA6YBUMGAfHSEPV0PetiB6zcLJKwVS3VyvBthWKgSCVsFwq8HdCyszhZcbg8XZQTfw9h4jjgAFlyFwHHNlZ9W0LXfSQGcABYQKx1r1wAAeEZQMCtgiw2DTrSn2Y5UCQsfwp1MfDiNFUg3LcV3LY8YDPKwKkiUMRw8TxLwQXIQO3Thb3oe8yk4J8cCCCJAFByX8RH-bBAMNYDN23VIyg6CCoJg0NUgw6gHyqWpcOwQj8CAA)

```python
import functools
from typing import Any, Callable, Concatenate, Self, overload


class Wrapper[**P, R]:
    def __init__(self, fn: Callable[P, R]) -> None:
        self.fn = fn

    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R:
        return self.fn(*args, **kwargs)

    @overload
    def __get__(self, obj: None, objtype: type | None = None) -> Self: ...

    @overload
    def __get__[**P, T](  # TypeVar "P" is already in use by an outer scope
        self: Wrapper[Concatenate[Any, P], T],
        obj: object,
        objtype: type | None = None,
    ) -> Callable[P, T]: ...

    def __get__(self, obj, objtype=None):  # type: ignore
        if obj is None:
            return self
        return functools.partial(self, obj)


def wrap[**P, R](fn: Callable[P, R]) -> Wrapper[P, R]:
    return Wrapper(fn)


class MyClass:
    @wrap
    def method(self, x: int) -> bool:
        return True

reveal_type(MyClass.method)    # Wrapper[(self: MyClass, x: int), bool]  ✅
reveal_type(MyClass().method)  # (x: int) -> bool  ✅

```

## Workaround

Use fresh module-level `ParamSpec`/`TypeVar` declarations (or 3.12+ inline type parameters) in the `__get__` overload instead of reusing the class-level type parameters.

## Environment

- Pyright version: 1.1.411
- Python version: 3.13
